### PR TITLE
drivers/spi: Add support for FPGA iCE40 bitstream loading.

### PR DESCRIPTION
**Pull request by janouja**
## Summary
drivers/spi: Add support for FPGA iCE40 bitstream loading.
Allows loading of bitstream directly from filesystem, either using ioctl or directly by using e.g. "cp example_bitstream.bin /dev/ice40-0"

arch and board esp32c3-legacy: Add optional iCE40 FPGA loading support 

## Impact
It should have no impact when CONFIG_SPI_ICE40 is not set (which is default)

## Testing
Tested on ICE-V-Wireless board.
https://github.com/ICE-V-Wireless/ICE-V-Wireless

### CODE REVIEW

File: drivers/spi/ice40.c [hunk 1]
```diff
@@ -0,0 +1,411 @@
...
+
+  if (!dev->in_progress)
+    {
+      _err ("ERROR: FPGA not initialized\n");
+      return -EINVAL;
+    }
+
+  dev->ops->select (dev, false);
+
+  for (size_t i = 0; i < 20; i++)
```
Comment (acassis) [inline drivers/spi/ice40.c:268]: Instead using a hard-coded value 20, please create a macro to indicate the value meaning.

**Comment by janouja** [comment]
Thanks for your review, I hope I fixed all your comments so please check again so it can be accepted to master.

File: include/nuttx/fs/ioctl.h [hunk 2]
```diff
@@ -693,6 +694,10 @@
 
 #define _BOARDIOCVALID(c) (_IOC_TYPE(c)==_BOARDBASE)
 #define _BOARDIOC(nr)     _IOC(_BOARDBASE,nr)
```
Comment (xiaoxiang781216) [inline include/nuttx/fs/ioctl.h:696]: add blnak line

File: drivers/spi/ice40.c [hunk 3]
```diff
@@ -0,0 +1,411 @@
...
+  int ret = OK;
+
+  switch (cmd)
+    {
+    case FPGAIOC_WRITE_INIT:
+      ret = ice40_init_fpga(dev);
+      break;
+
+    case FPGAIOC_WRITE:
+      ret = ice40_writeblk(dev, (FAR const char *)arg, sizeof (arg));
```
Comment (xiaoxiang781216) [inline drivers/spi/ice40.c:367]: sizeof(arg)

File: include/nuttx/spi/ice40.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+ *
+ ****************************************************************************/
+
+struct ice40_dev_s;
+
+struct ice40_ops_s
+{
+  CODE void(*reset) (FAR struct ice40_dev_s *dev, FAR bool reset);
+  CODE void(*select) (FAR struct ice40_dev_s *dev, FAR bool select);
+  CODE bool(*get_status) (FAR struct ice40_dev_s *dev);
```
Comment (xiaoxiang781216) [inline include/nuttx/spi/ice40.h:78]: remove space before (

File: include/nuttx/spi/ice40.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+#define ICE40_SPI_MODE (SPIDEV_MODE0) /* SPI Mode 0: CPOL=0,CPHA=0 */
+
+#define ICE40_SPI_FINAL_CLK_CYCLES 160
+
+#define ICE_SPI_MAX_XFER 4096
+
+#define FPGAIOC_WRITE_INIT _FPGACFGIOC (0x0001)
+#define FPGAIOC_WRITE _FPGACFGIOC (0x0002)
+#define FPGAIOC_WRITE_COMPLETE _FPGACFGIOC (0x0003)
```
Comment (xiaoxiang781216) [inline include/nuttx/spi/ice40.h:58]: remove space after _FPGACFGIOC

**Comment by acassis** [comment]
ping @janouja

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by acassis — APPROVED** [review]

**Comment by acassis** [comment]
Thank you very much @janouja !!!

**Pull request merged**